### DISCUSSION TITLE
How to Preventing Overlapping of Nodes When Nodes are Dragged or Resized

### DISCUSSION THREAD

**Discussion by osama2kabdullah** [xyflow/xyflow#2870] [Q&A]
I need to prevent nodes from overlapping when their size changes. Is there a way to automatically update the positions of the nodes when their size changes, so that they don't overlap with each other? I'd like the nodes to move away from each other when their size increases, and move closer to each other when their size decreases.

This is my component

```jsx
import { useCallback, useState } from "react";
import ReactFlow, {
  addEdge,
  applyEdgeChanges,
  applyNodeChanges,
} from "reactflow";
import "reactflow/dist/style.css";
import CutomNode from "./CutomNode.jsx";
import { TableTemplate } from "../Table/TableTemplate.js";

const rfStyle = {
  backgroundColor: "#B8CEFF",
};

const initialNodes = [
  {
    id: "node-1",
    type: "textUpdater",
    position: { x: -100, y: 0 },
    data: new TableTemplate(0, "My emplyee", 3),
  },
  {
    id: "node-2",
    type: "textUpdater",
    position: { x: 10, y: 300 },
    data: new TableTemplate(1, "My academi", 2),
  },
  {
    id: "node-3",
    type: "textUpdater",
    position: { x: 300, y: 20 },
    data: new TableTemplate(2, "My salary", 1),
  },
];
// we define the nodeTypes outside of the component to prevent re-renderings
const nodeTypes = { textUpdater: CutomNode };

function Flow() {
  const [nodes, setNodes] = useState(initialNodes);
  const [edges, setEdges] = useState([]);

  const onNodesChange = useCallback(
    (changes) => setNodes((nds) => applyNodeChanges(changes, nds)),
    [setNodes]
  );
  const onEdgesChange = useCallback(
    (changes) => setEdges((eds) => applyEdgeChanges(changes, eds)),
    [setEdges]
  );
  const onConnect = useCallback(
    (connection) => setEdges((eds) => addEdge(connection, eds)),
    [setEdges]
  );

  return (
    <ReactFlow
      nodes={nodes}
      edges={edges}
      onNodesChange={onNodesChange}
      onEdgesChange={onEdgesChange}
      onConnect={onConnect}
      nodeTypes={nodeTypes}
      snapToGrid={true}
      snapGrid={[500, 200]}
      fitView
      style={rfStyle}
    />
  );
}

export default Flow;
```

This is `CustomNode` component

```jsx
import Table from "../Table/Table";

export default function CutomNode({ data, isConnectable }) {
  return <Table data={data} />; //this component is indivital node
}
```
### Visualization of my code snippet. Here nodes are overlap one to another. But I don't want this.
[ncd-visual.webm](https://user-images.githubusercontent.com/109265357/221740577-1aec5b1f-d82d-444d-b4c8-418c07098db0.webm)

**Comment by connect-bilal** [comment]
You find out any answer? Because I'm looking same

**Reply by osama2kabdullah** [reply]
No. I did not found any answer yet.

**Reply by KriticaCelito** [reply]
Hii, did you find any answer I just want to shift my edge whenver it finds any node on the way..so that the edges don't overlap the node. Though I'm using dagre but unable to figure out

**Comment by dsoklic** [comment]
The Flow component has the `onNodeDragStop` callback, and every `NodeResizer` has a `onResizeEnd` callback. You can use those to figure out if nodes are overlapping and revert to a position and size where they don't overlap.

**Reply by mbehrisch** [reply]
I am actually also interested in this problem (my use case is only for dragging)

**Comment by akashbhadouria** [comment]
### This is a working example to understand how you will overcome this issue.

**When a node is being dragged and overlaps with another node (detected via getIntersectingNodes(node)), we can:

Remember the node’s original position when dragging starts (onNodeDragStart).

Continuously check for intersections while dragging (onNodeDrag).

If overlap detected → revert node back to original position (setNodes() inside onNodeDragStop).

This ensures a node snaps back to its original location whenever it tries to overlap.**

`import React, { useCallback, useRef, type MouseEvent } from 'react';
import {
  ReactFlow,
  Background,
  Controls,
  ReactFlowProvider,
  useReactFlow,
  useNodesState,
  type Edge,
  type Node,
} from '@xyflow/react';
import '@xyflow/react/dist/style.css';

const initialNodes: Node[] = [
  {
    id: '1',
    data: { label: 'Node 1' },
    position: { x: 0, y: 0 },
    style: { width: 200, height: 100 },
  },
  {
    id: '2',
    data: { label: 'Node 2' },
    position: { x: 0, y: 150 },
  },
  {
    id: '3',
    data: { label: 'Node 3' },
    position: { x: 250, y: 0 },
  },
  {
    id: '4',
    data: { label: 'Node 4' },
    position: { x: 350, y: 150 },
    style: { width: 50, height: 50 },
  },
];

const initialEdges: Edge[] = [];

const BasicFlow = () => {
  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
  const { getIntersectingNodes } = useReactFlow();

  // store original positions when drag starts
  const originalPositions = useRef<Record<string, { x: number; y: number }>>({});

  const onNodeDragStart = useCallback((_evt: MouseEvent, node: Node) => {
    originalPositions.current[node.id] = { ...node.position };
  }, []);

  const onNodeDrag = useCallback(
    (_evt: MouseEvent, node: Node) => {
      const intersections = getIntersectingNodes(node);
      console.log(
        `%c📍 Dragging Node ${node.id}:`,
        'color: cyan; font-weight: bold;',
        node.position
      );

      if (intersections.length > 0) {
        console.log(
          `%c⚠️ Overlapping with: ${intersections.map((n) => n.id).join(', ')}`,
          'color: orange; font-weight: bold;'
        );
      }

      // highlight intersecting nodes
      setNodes((ns) =>
        ns.map((n) => ({
          ...n,
          className: intersections.some((i) => i.id === n.id) ? 'highlight' : '',
        }))
      );
    },
    [getIntersectingNodes, setNodes]
  );

  const onNodeDragStop = useCallback(
    (_evt: MouseEvent, node: Node) => {
      const intersections = getIntersectingNodes(node);
      if (intersections.length > 0) {
        console.log(
          `%c🚫 Node ${node.id} overlapped — reverting to original position.`,
          'color: red; font-weight: bold;'
        );

        // revert to original position
        setNodes((ns) =>
          ns.map((n) =>
            n.id === node.id
              ? {
                  ...n,
                  position: originalPositions.current[node.id] || n.position,
                }
              : n
          )
        );
      }

      // clear highlight
      setNodes((ns) => ns.map((n) => ({ ...n, className: '' })));
    },
    [getIntersectingNodes, setNodes]
  );

  return (
    <ReactFlow
      nodes={nodes}
      edges={initialEdges}
      onNodesChange={onNodesChange}
      onNodeDragStart={onNodeDragStart}
      onNodeDrag={onNodeDrag}
      onNodeDragStop={onNodeDragStop}
      className="intersection-flow"
      minZoom={0.2}
      maxZoom={4}
      fitView
      selectNodesOnDrag={false}
    >
      <Background />
      <Controls />
    </ReactFlow>
  );
};

export default function App() {
  return (
    <ReactFlowProvider>
      <BasicFlow />
    </ReactFlowProvider>
  );
}
`

Reference - #https://reactflow.dev/examples/nodes/intersections
